nimbus/src/enrollment.rs

```diff
@@ -179,55 +525,117 @@ pub fn reset_enrollment(
-    // they don't the enrollment will just be removed next time we refresh.
+pub fn opt_in_with_branch(
+    db: &Database,
+    experiment_slug: &str,
+    branch: &str,
+) -> Result<Vec<EnrollmentChangeEvent>> {
+    let mut writer = db.write()?;
+    let exp_store = db.get_store(StoreId::Experiments);
+    let exp = match exp_store.get::<Experiment>(&writer, experiment_slug)? {
+        None => {
+            // XXX - do we want specific errors for this kind of thing?
+            log::warn!("No such experiment '{}'", experiment_slug);
+            return Ok(vec![]);
+        }
+        Some(e) => e,
+    };
+    if !exp
+        .branches
+        .iter()
+        .map(|b| &b.slug)
+        .any(|branch_slug| branch == branch_slug)
+    {
+        // XXX - do we want specific errors for this kind of thing?
+        log::warn!(
+            "No such branch '{}' for experiment {}",
+            branch,
+            experiment_slug
+        );
+        return Ok(vec![]);
+    }
     let enrollment = ExperimentEnrollment {
         slug: experiment_slug.to_string(),
-        status: EnrollmentStatus::Enrolled {
-            reason: EnrolledReason::OptIn,
-            branch: branch.to_string(),
-        },
+        status: EnrollmentStatus::enrolled(EnrolledReason::OptIn, branch),
+    };
+    let enrollment_id = if let EnrollmentStatus::Enrolled { enrollment_id, .. } = &enrollment.status
+    {
+        enrollment_id
+    } else {
+        unreachable!(); // This is really dumb.
     };
-    let mut writer = db.write()?;
     db.get_store(StoreId::Enrollments)
         .put(&mut writer, experiment_slug, &enrollment)?;
     writer.commit()?;
-    Ok(())
+    Ok(vec![EnrollmentChangeEvent::new(
+        &enrollment.slug,
+        enrollment_id,
+        branch,
+        None,
+        EnrollmentChangeEventType::Enrollment,
+    )])
 }
 
-pub fn opt_out(db: &Database, experiment_slug: &str) -> Result<()> {
-    // As above - check experiment exists?
+pub fn opt_out(db: &Database, experiment_slug: &str) -> Result<Vec<EnrollmentChangeEvent>> {
+    let mut writer = db.write()?;
+    let exp_store = db.get_store(StoreId::Experiments);
+    let _exp = match exp_store.get::<Experiment>(&writer, experiment_slug)? {
+        None => {
+            // XXX - do we want specific errors for this kind of thing?
+            log::warn!("No such experiment '{}'", experiment_slug);
+            return Ok(vec![]);
+        }
+        Some(e) => e,
+    };
     let enrollment = ExperimentEnrollment {
         slug: experiment_slug.to_string(),
-        status: EnrollmentStatus::OptedOut,
+        status: EnrollmentStatus::Disqualified {
+            reason: DisqualifiedReason::OptOut,
+            // TODO: We should use the branch we're enrolled in if we had an enrollment
+            // But what do we do if we did not? I assume we want to support the "opt out
+            // before enrollment ever happens use case". Should we just make "branch" optional?
```

Comment:
This currently does not happen in JS Desktop Nimbus. We check the `bucketConfig`/sampling to see if the user qualifies and then call `enroll` which [we are certain will succeed and choose a branch](https://searchfox.org/mozilla-central/rev/66547980e8e8ca583473c74f207cae5bac1ed541/toolkit/components/messaging-system/experiments/ExperimentManager.jsm#120-123).
And then on unenrollment we send the branch [in the event](https://searchfox.org/mozilla-central/rev/66547980e8e8ca583473c74f207cae5bac1ed541/toolkit/components/messaging-system/experiments/ExperimentManager.jsm#301).